netaddr.go

```diff
@@ -846,3 +871,422 @@ func (p *IPPrefix) UnmarshalText(text []byte) error {
+	// From is the starting IP address in the range.
+	// It must have the same address family (IPv4 vs IPv6) as To
+	// and be less than or equal to To.
+	// From is an inclusive bound; it is included in the set.
+	From IP
+
+	// To is the ending IP address in the range.
+	// It must have the same address family (IPv4 vs IPv6) as From
+	// and be greater than or equal to From.
+	// To is an inclusive bound; it is included in the set.
+	To IP
+}
+
+// Valid reports whether r.From and r.To are both non-zero and obey
+// the documented requirements: address families match, and From is
+// less than or equal to To.
+func (r IPRange) Valid() bool {
+	return !r.From.IsZero() && !r.To.IsZero() &&
+		r.From.Is4() == r.To.Is4() &&
+		!r.To.Less(r.From)
+}
+
+// ip16 represents a mutable IP address, either IPv4 (in IPv6-mapped
+// form) or IPv6.
+type ip16 [16]byte
+
+// bitSet reports whether the given bit in the address is set.
+// (bit 0 is the most significant bit in ip[0]; bit 127 is last)
+func (ip ip16) bitSet(bit uint8) bool {
+	i, s := bit/8, 7-(bit%8)
+	return ip[i]&(1<<s) != 0
+}
+
+func (ip *ip16) set(bit uint8) {
+	i, s := bit/8, 7-(bit%8)
+	ip[i] |= 1 << s
+}
+
+func (ip *ip16) clear(bit uint8) {
+	i, s := bit/8, 7-(bit%8)
+	ip[i] &^= 1 << s
+}
+
+// lastWithBitZero returns a copy of ip with the given bit
+// cleared and the following all set.
+func (ip ip16) lastWithBitZero(bit uint8) ip16 {
+	ip.clear(bit)
+	for ; bit < 128; bit++ {
+		ip.set(bit)
+	}
+	return ip
+}
+
+// firstWithBitOne returns a copy of ip with the given bit
+// set and the following all cleared.
+func (ip ip16) firstWithBitOne(bit uint8) ip16 {
+	ip.set(bit)
+	for ; bit < 128; bit++ {
+		ip.clear(bit)
+	}
+	return ip
+}
+
+// prefixMaker returns a address-family-corrected IPPrefix from ip16 and bits,
+// where the input bits is always in the IPv6-mapped form for IPv4 addresses.
+type prefixMaker func(ip16 ip16, bits uint8) IPPrefix
+
+// Prefixes returns the set of IPPrefix entries that covers r.
+//
+// If either of r's bounds are invalid, in the wrong order, or if
+// they're of different address families, then Prefixes returns nil.
+func (r IPRange) Prefixes() []IPPrefix {
+	if !r.Valid() {
+		return nil
+	}
+	var makePrefix prefixMaker
+	if r.From.Is4() {
+		makePrefix = func(ip16 ip16, bits uint8) IPPrefix {
+			return IPPrefix{IPFrom16([16]byte(ip16)), bits - 12*8}
+		}
+	} else {
+		makePrefix = func(ip16 ip16, bits uint8) IPPrefix {
+			return IPPrefix{IPv6Raw([16]byte(ip16)), bits}
+		}
+	}
+	a16, b16 := ip16(r.From.As16()), ip16(r.To.As16())
+	return appendRangePrefixes(nil, makePrefix, a16, b16)
+}
+
+func appendRangePrefixes(dst []IPPrefix, makePrefix prefixMaker, a16, b16 ip16) []IPPrefix {
+	common := uint8(0)
+	for common < 128 && a16.bitSet(common) == b16.bitSet(common) {
+		common++
+	}
+	// See whether a16 and b16, after their common shared bits, end
+	// in all zero bits or all one bits, respectively.
+	aAllZero, bAllSet := true, true
```

Comment:
This can be a single bool; you can combine the following two loops.

```go
allZeroOne := true
for i := common; i < 128; i++ {
  if a16.bitSet(i) || !b16.bitSet(i) {
    allZeroOne = false
    break
  }
}
```